### PR TITLE
net/gcoap: Confirmable Observe notifications

### PR DESCRIPTION
**Pull request by kb2ma**
This PR implements the fourth step in the confirmable messaging plan in #7192, and depends on #7337. It allows for an Observe notification to be sent confirmably. Also, the server may deregister a client from further notifications based on the response. The client is deregistered if it responds to the notification with a reset (RST) message, or if the server does not receive a response from the client. This PR also includes refactoring of a couple of utility methods to support the implementation.

### CODE REVIEW

**Comment by smlng** [comment]
#7337 is merged, please rebase

**Comment by kb2ma** [comment]
Rebased.

File: sys/include/net/gcoap.h [hunk 1]
```diff
@@ -158,8 +158,9 @@
  * GCOAP_OBS_VALUE_WIDTH.
  *
  * To cancel a notification, the server expects to receive a GET request with
- * the Observe option value set to 1. The server does not support cancellation
- * via a reset (RST) response to a non-confirmable notification.
+ * the Observe option value set to 1. A confirmable notification also is
+ * cancelled by a reset (RST) response from the client or by the absence of an
+ * ACK response.
```
Comment (miri64) [inline sys/include/net/gcoap.h]: Is this according to the standard? What if the ACK (or the CON for that matter) just get's lost?

**Comment by miri64** [comment]
Needs a rebase.

File: sys/include/net/gcoap.h (hunk 1 above)
Comment (kb2ma) [reply sys/include/net/gcoap.h]: Yes, according to RFC 7641. There is a short summary at the end of Sec 1.2:

> When the client deregisters, rejects a notification, or the transmission of a notification
> times out after several transmission attempts, the client is considered no longer interested
> in the resource and is removed by the server from the list of observers.

See also Sec 4.5 specifically on retransmissions. This removal of a client is a MUST (in the RFC 2119 sense).

Observe is meant to be a best-effort attempt to notify the client. It's not a guaranteed delivery mechanism. However, the server can include a Max-Age option in the notification so the client can decide to re-register for the resource if it hasn't heard from the server.

The code currently does not include a Max-Age option. See #8831 for the plan to make it easy to add arbitrary options to a message.

**Comment by kb2ma** [comment]
Rebase looks good.

File: sys/include/net/gcoap.h (hunk 1 above)
Comment (miri64) [reply sys/include/net/gcoap.h]: >> […] or the transmission of a notification times out after several transmission attempts, […]

How many retransmission attempts are there in this implementation?

Comment (kb2ma) [reply sys/include/net/gcoap.h]: Here is the quote from [sec. 4.5](https://tools.ietf.org/html/rfc7641#section-4.5):

> The client's acknowledgement of a confirmable notification signals
> that the client is interested in receiving further notifications.  If
> a client rejects a confirmable or non-confirmable notification with a
> Reset message, or if the last attempt to retransmit a confirmable
> notification times out, then the client is considered no longer
> interested and the server MUST remove the associated entry from the
> list of observers.

So, the number of retransmissions is governed by the confirmable retransmission count in the implementation. nanocoap sets COAP_MAX_RETRANSMIT to 4, so 5 transmission attempts are made altogether, using exponential backoff (2, 4, 8, 16 seconds).

You also might be interested to read at the end of sec. 4.5 about non-confirmable transmission. If notifications are sent non-confirmably, at least once a day the implementation MUST send a transmission confirmably. So, the spec uses the confirmable retransmission limit to eliminate Observe clients no longer interested in the resource.

This requirement to send a message confirmably at least once a day is not enforced by gcoap.

**Comment by miri64** [comment]
Sorry, this somehow got lost. @kb2ma can you rebase again, please?

**Comment by kb2ma** [comment]
Thanks for circling back on this one, @miri64. It's been quite a while since the original work, so I'd like to review this myself as well. However, my head is deeply in #10732 at the moment. Let me start generating PRs for that, and then I'll work on this one.

**Pull request closed without merging**